File: level2/GA/main.cpp
#include <iostream>
#include "CppMaze.h"

using namespace std;

typedef pair<int, int> POINT_;
typedef vector<POINT_> PATH_;

/* 定义常量 */
const int ROW = 10;
const int COLUMN = 10;
const int LOOPPERCENT = 50;
char DIRECTIONS[4] = {'N', 'S', 'W', 'E'};
const int INDIVIDUALSNUM = 50;
const double HERITABILITY = 0.2;
const int ITERATENUM = 50;

/* 定义变量 */
POINT_ LastPoint = make_pair(1, 1);
PATH_ OptimalPath;
map<POINT_, double> AlterTropism;
map<POINT_, double> Tropism;
MAZE Maze = MAZE(ROW, COLUMN);
PATH_ Path;

int Manhattan(int x1, int y1, int x2, int y2);
double Euclid(int x1, int y1, int x2, int y2);
POINT_ Roulette(POINT_ EventList[], double ProList[]);
void MazeInit();
void TropismInit();
void WaneTropism(int Row, int Column);
void AddTropism(PATH_ PathList);
void UpdateTropism();
bool SingleMovement(int Row, int Column);
void GenerationMovement(int Row, int Column);

int main() {
    MazeInit();
    TropismInit();

    int IterateNum = 0;
    clock_t begin = clock();
    while(IterateNum <= ITERATENUM){
        GenerationMovement(1, 1);
        IterateNum += 1;
    }
    clock_t end = clock();
    Maze.DrawConsoleMaze();
    HANDLE hOnt = GetStdHandle(STD_OUTPUT_HANDLE);
    COORD Pos;
    CONSOLE_SCREEN_BUFFER_INFO screenInfo;
    GetConsoleScreenBufferInfo(hOnt, &screenInfo);
    Pos.X = Pos.Y = 1;
    SetConsoleCursorPosition(hOnt, Pos);
    cout << "O";
    for(auto iter : OptimalPath) {
        if(Pos.X + 2 == 2 * iter.second - 1)
            Pos.X++;
        else if(Pos.X - 2 == 2 * iter.second - 1)
            Pos.X--;
        else if(Pos.Y + 2 == 2 * iter.first - 1)
            Pos.Y++;
        else if(Pos.Y - 2 == 2 * iter.first - 1)
            Pos.Y--;
        SetConsoleCursorPosition(hOnt, Pos);
        cout << "x";
        Pos.X = 2 * iter.second - 1;
        Pos.Y = 2 * iter.first - 1;
        SetConsoleCursorPosition(hOnt, Pos);
        if(Pos.Y != 2 * ROW - 1 || Pos.X != 2 * COLUMN - 1)
            cout << "x";
    }
    Pos.Y = 2 * ROW + 1;
    Pos.X = 0;
    SetConsoleCursorPosition(hOnt, Pos);
    cout << "Time used:" << (double)(end - begin) / CLOCKS_PER_SEC << "s";
    getchar();
    return 0;
}

int Manhattan(int x1, int y1, int x2, int y2) {
    return abs(x1 - x2) + abs(y1 - y2);
}

double Euclid(int x1, int y1, int x2, int y2) {
    return sqrt(abs(pow(x1, 2) - pow(x2, 2)) + abs(pow(y1, 2) - pow(y2, 2)));
}

POINT_ Roulette(PATH_ EventList, vector<double> ProList) {
    for(auto iter = ProList.begin() + 1; iter != ProList.end(); iter++)
        *iter += *(iter - 1);
    auto iter = ProList.end();
    iter--;
    *iter = 1;
    random_device RANDOM;
    double RandomNum = (RANDOM() % 100 + 1) / 100.0;
    for(auto iter = ProList.begin(); iter != ProList.end(); iter++)
        if(RandomNum <= *iter)
            return EventList[iter - ProList.begin()];
}

void MazeInit() {
    Maze.CreateMaze(ROW, COLUMN, LOOPPERCENT);
}

void TropismInit() {
    Tropism.clear();
    for(int i = 1; i <= ROW; i++)
        for(int j = 1; j <= COLUMN; j++) {
            Tropism.insert(make_pair(make_pair(i, j), Manhattan(1, 1, i, j) + pow(Euclid(1, 1, i + 1, j + 1), 0.2)));
            AlterTropism.insert(make_pair(make_pair(i, j), 0.0));
        }
}

void WaneTropism(int Row, int Column) {
    Tropism[make_pair(Row, Column)] = 0.1;
}

void AddTropism(PATH_ PathList) {
    for(auto iter : PathList)
        AlterTropism[iter] += (ROW + COLUMN) / PathList.size() + Tropism[iter];
}

void UpdateTropism() {
    for(int i = 1; i <= ROW; i++)
        for(int j = 1; j <= COLUMN; j++) {
            Tropism[make_pair(i, j)] += AlterTropism[make_pair(i, j)];
            AlterTropism.insert(make_pair(make_pair(i, j), 0.0));
        }
}

bool SingleMovement(int Row, int Column) {
    PATH_ DirectionsList;
    vector<double> ProList;
    double TotalTropism = 0;

    WaneTropism(Row, Column);
    
    for(int i = 0; i < 4; i++)
        if(Maze.MazeMap[make_pair(Row, Column)][DIRECTIONS[i]])
            switch(DIRECTIONS[i]) {
                case 'N': {
                    if(LastPoint != make_pair(Row - 1, Column)) {
                        DirectionsList.emplace_back(Row - 1, Column);
                        TotalTropism += Tropism[make_pair(Row - 1, Column)];
                    }
                    break;
                }
                case 'S': {
                    if(LastPoint != make_pair(Row + 1, Column)) {
                        DirectionsList.emplace_back(Row + 1, Column);
                        TotalTropism += Tropism[make_pair(Row + 1, Column)];
                    }
                    break;
                }
                case 'W': {
                    if(LastPoint != make_pair(Row, Column - 1)) {
                        DirectionsList.emplace_back(Row, Column - 1);
                        TotalTropism += Tropism[make_pair(Row, Column - 1)];
                    }
                    break;
                }
                case 'E': {
                    if(LastPoint != make_pair(Row, Column + 1)) {
                        DirectionsList.emplace_back(Row, Column + 1);
                        TotalTropism += Tropism[make_pair(Row, Column + 1)];
                    }
                    break;
                }
            }

    POINT_ LastBifurcate;
    char LastDirection;

    if(!DirectionsList.empty()) {
        for(auto iter : DirectionsList)
            ProList.push_back(Tropism[iter] / TotalTropism);
        Path.push_back(Roulette(DirectionsList, ProList));
        LastPoint = make_pair(Row, Column);
        if(DirectionsList.size() > 1) {
            LastBifurcate = make_pair(Row, Column);
            if(Path[Path.size() - 1] == make_pair(Row - 1, Column))
                LastDirection = 'N';
            else if(Path[Path.size() - 1] == make_pair(Row + 1, Column))
                LastDirection = 'S';
            else if(Path[Path.size() - 1] == make_pair(Row, Column - 1))
                LastDirection = 'W';
            else if(Path[Path.size() - 1] == make_pair(Row, Column - 1))
                LastDirection = 'E';
        }
        return true;
    }

    Maze.MazeMap[LastBifurcate][LastDirection] = 0;
    return false;
}

void GenerationMovement(int Row, int Column) {
    map<int, PATH_> Paths;
    int IndividualNum = 0;
    if(!OptimalPath.empty())
        Paths.insert(make_pair(OptimalPath.size(), OptimalPath));
    while(IndividualNum <= INDIVIDUALSNUM) {
        Path.clear();
        int Step = 0;
        int Row = 1;
        int Column = 1;
        while(Step <= ROW * COLUMN) {
            if(SingleMovement(Row, Column)) {
                Row = (Path.end() - 1) -> first;
                Column = (Path.end() - 1) -> second;
                if(Row == ROW && Column == COLUMN) {
                    if(Paths.size() <= INDIVIDUALSNUM * HERITABILITY)
                        Paths.insert(make_pair(Path.size(), Path));
                    else {
                        auto iter = Paths.end();
                        iter--;
                        if (Path.size() < iter -> first) {
                            Paths.erase(iter);
                            Paths.insert(make_pair(Path.size(), Path));
                        }
                    }
                }
                Step++;
            }
            else
                break;
        }

        if(!Paths.empty()) {
            OptimalPath.clear();
            for(auto iter : Paths.begin() -> second)
                OptimalPath.push_back(iter);
        }
        for(const auto& iter : Paths)
            AddTropism(iter.second);
        UpdateTropism();
        IndividualNum += 1;
    }
}

File: level2/GA/main.py
from pyamaze import maze, COLOR, agent
from numpy import random
from math import *
import time

""" 定义常量 """
ROW = 30
COLUMN = 30
LOOPPERCENT = 50
DIRECTIONS = ['N', 'S', 'W', 'E']
INDIVIDUALSNUM = 50  # 一代中个体数目
HERITABILITY = 0.2  # 遗传个体比率
ITERATENUM = 50  # 迭代次数

""" 定义变量 """
LastPoint = [1, 1]
OptimalPath = []  # 存储当前最优路径
AlterTropism = {}  # 趋向度改变值

""" 计算曼哈顿距离 """
def Manhattan(x1, y1, x2, y2):
    return abs(x1 - x2) + abs(y1 - y2)

""" 计算欧几里得距离 """
def Euclid(x1, y1, x2, y2):
    return sqrt(abs(x1 ** 2 - x2 ** 2) + abs(y1 ** 2 - y2 ** 2))

""" 轮盘赌计算概率 """
def Roulette(EventList, ProList):
    EventListIndex = range(len(EventList))
    return EventList[random.choice(EventListIndex, p = ProList)]

""" 初始化迷宫 """
def MazeInit():
    global Maze, Agent
    Maze = maze(ROW, COLUMN)
    Maze.CreateMaze(ROW, COLUMN, theme = COLOR.dark, loopPercent = LOOPPERCENT)
    Agent = agent(Maze, x = 1, y = 1, goal = (ROW, COLUMN), footprints = True)

""" 初始化趋向度（T） """
def TropismInit():
    # 定义趋向度（T）公式为(Manhattan ^ 2) * Euclid
    global Tropism
    Tropism = {}
    for i in range(ROW):
        for j in range(COLUMN):
            Tropism.update({(i + 1, j + 1): Manhattan(1, 1, i + 1, j + 1) + Euclid(1, 1, i + 1, j + 1) ** 0.1})
            AlterTropism.update({(i + 1, j + 1): 0})


""" 减弱趋向度（T） """
def WaneTropism(Row, Column):
    Tropism[(Row, Column)] -= Tropism[(Row, Column)] - 0.01


""" 增强趋向度（T） """
def AddTropism(PathList):
    for Unit in PathList:
        AlterTropism[Unit] += (ROW + COLUMN) / len(PathList) + Tropism[Unit]


""" 更新趋向度（T） """
def UpdateTropism():
    for i in range(ROW):
        for j in range(COLUMN):
            Tropism[(i + 1, j + 1)] += AlterTropism[(i + 1, j + 1)]
            AlterTropism.update({(i + 1, j + 1): 0})

""" 判断生物移动方向 """
def SingleMovement(Row, Column):
    DirectionsList = []
    ProList = []
    TotalTropism = 0

    WaneTropism(Row, Column)

    for Direction in DIRECTIONS:
        if Maze.maze_map[(Row, Column)][Direction] == 1:
            match Direction:
                case 'N':
                    if tuple(LastPoint) != (Row - 1, Column):
                        DirectionsList.append((Row - 1, Column))
                        TotalTropism += Tropism[(Row - 1, Column)]
                case 'S':
                    if tuple(LastPoint) != (Row + 1, Column):
                        DirectionsList.append((Row + 1, Column))
                        TotalTropism += Tropism[(Row + 1, Column)]
                case 'W':
                    if tuple(LastPoint) != (Row, Column - 1):
                        DirectionsList.append((Row, Column - 1))
                        TotalTropism += Tropism[(Row, Column - 1)]
                case 'E':
                    if tuple(LastPoint) != (Row, Column + 1):
                        DirectionsList.append((Row, Column + 1))
                        TotalTropism += Tropism[(Row, Column + 1)]

    if len(DirectionsList) != 0:
        for Direction in DirectionsList:
            ProList.append(Tropism[Direction] / TotalTropism)
        Path.append(Roulette(DirectionsList, ProList))
        LastPoint[0] = Row
        LastPoint[1] = Column
        return True

    return False

""" 生物一代中的移动 """
def GenerationMovement(Row, Column):
    Paths = {}
    IndividualNum = 0
    if OptimalPath:
        Paths.update({tuple(OptimalPath): len(OptimalPath)})
    while IndividualNum <= INDIVIDUALSNUM:
        global Path
        Path = []
        Step = 0
        Row = 1
        Column = 1
        while Step <= ROW * COLUMN:
            if SingleMovement(Row, Column):
                Row = Path[len(Path) - 1][0]
                Column = Path[len(Path) - 1][1]
                if Row == ROW and Column == COLUMN:
                    if len(Paths) <= INDIVIDUALSNUM * HERITABILITY:
                        Paths.update({tuple(Path): len(Path)})
                    else:
                        if len(Path) < Paths[list(Paths.keys())[-1]]:
                            Paths.pop(list(Paths.keys())[-1])
                            Paths.update({tuple(Path): len(Path)})
                    PathsList = sorted(Paths.items(), key=lambda x: x[-1])
                    Paths.clear()
                    for path in PathsList:
                        Paths.update({tuple(path[0]): path[1]})
                Step += 1
            else:
                break

        if len(list(Paths.keys())) != 0:
            OptimalPath.clear()
            for Unit in list(Paths.keys())[0]:
                OptimalPath.append(Unit)
        for path in Paths.keys():
            AddTropism(list(path))
        UpdateTropism()
        IndividualNum += 1

if __name__ == '__main__':
    MazeInit()
    TropismInit()
    T1 = time.time()
    IterateNum = 0
    while True:
        while IterateNum <= ITERATENUM:
            GenerationMovement(1, 1)
            IterateNum += 1
        if len(OptimalPath) == 0:
            IterateNum -= 3
        else:
            break
    T2 = time.time()
    print(T2 - T1)
    print(len(OptimalPath))
    Maze.tracePath({Agent: OptimalPath}, delay = 10)
    Maze.run()